odoo/models.py

```diff
@@ -6709,6 +6717,40 @@ def flush_recordset(self, fnames=None):
         if self.env.cache.has_dirty_fields(self, fields_):
             self._flush(fnames)
 
+    def _ensure_exists(self):
+        """ enusre records' existence in the current and all concurrent transactions """
+        # Lock referenced records by performing a dummy update to prevent concurrent deletion
+        ids = set(self._ids)
+        if not ids:
+            return
+        self.env.cr.execute(SQL("""
+            UPDATE %(table)s
+            SET id = id
+            WHERE id IN %(ids)s
+            """,
+            table = SQL.identifier(self._table),
+            ids = tuple(ids),
+        ))
```

Comment:
Multiple updates for a row are bad.  Instead of updating, you should simply lock the rows for update with
```sql
SELECT id FROM ...
WHERE id IN (...)
FOR UPDATE
```
You could even use `FOR UPDATE NO WAIT` to exit immediately if the given rows are locked already.

Reply:
I thought about this strategy by using `FOR KEY SHARE` when write/flush, and `FOR UPDATE NO WAIT` when unlink
but it doesn't work for concurrent DELETE

e.g. we have 2 translactions A, B
A.start ..............A.assign_m2o.......A.lock_row......A.commit(release lock)
.............B.start........................................................B.cleanup_m2o..............B.delete_row.......... B.commit

Since B starts after A, it cannot see the new relation added by A.assign_m2o and cannot clean it.
and the A.lock_row can only block B.delete_row before A.commit. But if A.commit before B.delete_row, B still can commit


As a solution, I move the force UPDATE in precomit hook, so one row will be updated by this SQL at most once